### PR TITLE
[WhyisQuek] iP

### PR DESCRIPTION
**Pull request by WhyisQuek**
## Timmy 

> "Your mind is for having ideas, not holding them" - David Allen [source](https://dansilvestre.com/productivity-quotes)

Timmy frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you need to do is,

1. download it from here
2. double-click it
3. add your tasks
4. let it manage tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

### CODE REVIEW

File: src/main/java/Timmy/TaskList.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
+package Timmy;
+
+import java.util.ArrayList;
+
+import Exceptions.TimmyTaskListOutOfBoundsException;
+
+/**
+ * Represents the list of tasks given by the user that has been stored in Timmy
```
Comment (jasonleexc) [inline src/main/java/Timmy/TaskList.java]: I like all the documentation heading comments, it makes the code very easy to follow!

File: src/main/java/Timmy/Storage.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            throw new TimmyFilerException();
+        }
+    }
+
+    /**
+     * Loads the saved list of tasks saved within the save file.
+     *
+     * @return the list of tasks that was saved within the save file.
+     */
+    public static ArrayList<Task> loadStorage() {
```
Comment (jasonleexc) [inline src/main/java/Timmy/Storage.java]: Good formatting of methods to be verbs!

File: src/test/java/Timmy/DeadlineTest.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package Timmy;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+
+public class DeadlineTest {
```
Comment (jasonleexc) [inline src/test/java/Timmy/DeadlineTest.java]: Perhaps could consider giving more tests for deadline, such as for invalid inputs to test the throwing of exceptions

File: src/test/java/Timmy/ToDoTest.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package Timmy;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+public class ToDoTest {
```
Comment (jasonleexc) [inline src/test/java/Timmy/ToDoTest.java]: similarly, you could consider testing invalid inputs, as well as other edge case inputs such as wrong formatting to test the throwing of exceptions

**Review by jasonleexc — COMMENTED** [review]
Well formatted with heading comments. LGTM! Just some small nits to pick.

File: src/main/java/Timmy/Deadline.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    protected static final DateTimeFormatter OUTPUT_DATE_FORMAT = DateTimeFormatter.ofPattern("MMM dd yyyy");
+    protected final LocalDate end;
+
+    /**
+     * Creates an incomplete task with a specified end date.
+     *
+     * @param desc  description of the event.
+     * @param end   end date of event, in d/M/yyyy format.
+     */
+    Deadline(String desc, String end) throws TimmyDateParsingException {
```
Comment (manishkharthik) [inline src/main/java/Timmy/Deadline.java]: End is slightly vague, could you have perhaps used a clearer noun like endDate or dueDate instead?

File: src/main/java/Timmy/Event.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+import Exceptions.TimmyDateParsingException;
+
+/**
+ * Represents a task with a start and end date.
+ */
+public class Event extends Task {
+    protected static final DateTimeFormatter INPUT_DATE_FORMAT = DateTimeFormatter.ofPattern("d/M/yyyy");
```
Comment (manishkharthik) [inline src/main/java/Timmy/Event.java]: I like how the fields have been named (snake upper case for static fields), and they are descriptive!

File: src/main/java/Timmy/Parser.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * Parses the parameters given for the event command. Parameters must
+     * follow the format [DESCRIPTION] /from [START_DATE] /to [END_DATE]
+     *
+     * @param   input Input parameters given by the user for the event command.
+     * @return  An array of 3 strings: {DESCRIPTION, START_DATE, END_DATE}
+     * @throws  TimmyInvalidParamException if the given input does not follow
+     *          the format specified above.
+     */
+    public static String[] parseEvent(String input) throws TimmyInvalidParamException {
```
Comment (manishkharthik) [inline src/main/java/Timmy/Parser.java]: I like the verbs used to describe the methods (parseMark, parseEvent and parseDeadline)

File: src/main/java/Timmy/Task.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    public String getStatusIcon() {
+        return (this.isDone ? "[X]" : "[ ]");
+    }
+
+    /**
+     * Returns the task as a formatted string with a configured status icon.
+     *
+     * @return the formatted string with a configured status icon.
+     */
+    public String toStringWithStatusIcon() {
```
Comment (manishkharthik) [inline src/main/java/Timmy/Task.java]: This method name is a little bit verbose, is there a way you can shorten it instead?

File: src/main/java/Timmy/Ui.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        }
+    }
+
+    /**
+     * Prints the message to be shown upon successfully adding a task.
+     *
+     * @param task the task that has been added.
+     * @param size the number of tasks in the list after successful addition.
+     */
+    public void showAddMessage(Task task, Integer size) {
```
Comment (manishkharthik) [inline src/main/java/Timmy/Ui.java]: I think size here actually refers to taskCount, could renaming the parameter in this method make it clearer?

**Review by manishkharthik — COMMENTED** [review]
Overall, the naming conventions follow that of the expectations. Well done!

File: src/main/java/Timmy/TaskList.java (hunk 1 above)
Comment (WhyisQuek) [reply src/main/java/Timmy/TaskList.java]: Thank you!

File: src/test/java/Timmy/DeadlineTest.java (hunk 3 above)
Comment (WhyisQuek) [reply src/test/java/Timmy/DeadlineTest.java]: You're right, I'll update the tests to be more comprehensive

File: src/main/java/Timmy/Deadline.java (hunk 5 above)
Comment (WhyisQuek) [reply src/main/java/Timmy/Deadline.java]: Will do!
